### DISCUSSION TITLE
Applying motion to the player/targeted entity?

### DISCUSSION THREAD

**Discussion by ProfLDZ** [FabricMC/fabric-api#2075] [Mod Dev Support] [answered]
I've been trying to apply motion to the player for over 2 hours, but nothing seems to work. All the examples I've looked at (projectile code) just seem to be using #setVelocity(), but when I do that it doesn't do anything. I just want to be able to apply motion via a command.

**Comment by Linguardium** [comment]
setVelocity and updateVelocity both work (in different ways) but must be called via the server not the client.

**Reply by ProfLDZ** [reply]
I'm still really new to Fabric modding, I'm not sure what you mean by calling it via the server. Like `source.getServer()`? And do you know how set/update Velocity differ?

**Reply by Linguardium** [reply]
https://fabricmc.net/wiki/tutorial:side

updateVelocity takes a speed (multiplier) and a movementInput (sideways, vertical, forward) Vec3D and applies the appropriate velocity based on those values.
setVelocity takes X, Y, Z velocity Vec3D directly

**Reply by ProfLDZ** [reply]
After reading the wiki you sent, I'm still confused about how I'd actually apply the velocity via the server? The closest thing I gathered from that wiki is that I _could_ send a packet to the server with velocity information (maybe; the only packet-related knowledge I have is from Spigot/Paper plugin development.)

**Reply by Linguardium** [reply]
any code you use to change velocity of an entity MUST be done on the server. where that code is depends on your usecase. if it is in a new custom projectile class, make sure you make a world.isClient check and only run the code if isClient returns false.

**Reply by ProfLDZ** [reply]
It's as a command that the player executes, so idk how it'd apply the velocity server-side

**Comment by warjort** [comment, marked as answer]
See this previous discussion which is largely a repeat of some your questions: https://github.com/FabricMC/fabric/discussions/1949

As to your specific requirement, just look at ````net.minecraft.server.command.TeleportCommand```` and change it to alter the player's velocity instead of doing a teleport.
It's also possible to ````ServerCommandSource.getPlayer()```` so the player is implicitly the player running the command, instead of specifying the player as a parameter (the target) like the teleport command does.

**Reply by ProfLDZ** [reply]
I looked at the teleport but I don't really see how making it set velocity instead of teleporting would do anything. It has `PlayerPositionLookS2CPacket` but that only seems to accept position-like parameters. There's another part farther down and it just uses `setVelocity` as well, which I want to use, but nothing happens when I use it (probably cause the player's executing the command?).

I'm already using `ServerCommandSource.getPlayer()` to mess with the player.

**Reply by warjort** [reply]
If you are just going to assert that you are doing it correctly without showing any code, error messages and/or debug statements that confirm you made the changes to the game state you expected (including to the correct player), there isn't much more we can do to help.
You should also check your logs/latest.log for errors/warnings.

As for your other comments about the teleport command:
The PlayerPositionLookS2CPacket is just used to get flags used for the ````requestTeleport()```` method.
The only use of velocity I can see is to remove any vertical velocity (mulitply y velocity by 0) after the teleport under certain circumstances.

**Reply by ProfLDZ** [reply]
I didn't provide code because it's literally `source.getPlayer().setVelocity` that I'm using and it not doing anything. There's no errors in the console, and I did debug messages checking the velocity before and after, and it is setting the velocity, it's just not doing anything.

**Reply by dvitski** [reply]
You have to also set `velocityDirty`/`velocityModified` to true (it's one of those two, I can't remember 😅)

**Reply by ProfLDZ** [reply]
I tried doing `velocityDirty` after I did `#setVelocity` but it didn't help, I'll try `velocityModified`

**Reply by warjort** [reply]
> You have to also set velocityDirty/velocityModified to true

I believe the dirty/modified flags just forces the update to be sent to the client immediately instead of being sent periodically according to the entity tracking config.
The actual movement is done during the entity tick on the server and has nothing to do with what the client knows.
The client knowing the velocity is just used to make the motion look more smooth than if it only uses "jerky" updates from the server.

As I said above, this question is currently unanswerable. It boils down to:
"I want to do X. I have done everything correctly but won't show you what I have done. Please help."
My guess is adding simple ````System.out.println()````s to relevant places will quickly show the correct game state is not being changed in the way expected/assumed.

**Reply by warjort** [reply]
````
I believe the dirty/modified flags just forces the update to be sent to the client immediately instead of being sent periodically according to the entity tracking config.
````
Actually, that might be the issue. 
I don't think a player tracks its own player entity. So it won't be receiving updates on the client when you change the velocity like it does for other players/entities.
The teleport command doesn't have this problem since it does network requests to tell the client what is going on.

So the solution would be to send a network request to the client telling it the velocity has changed.
You can probably just reuse ````net.minecraft.network.packet.s2c.play.EntityVelocityUpdateS2CPacket````?
